subprojects/docs/src/docs/userguide/javaPlugin.adoc

```diff
@@ -481,37 +481,42 @@ By default, the Java compiler runs in the Gradle process. Setting `options.fork`
 
 Starting with Gradle 2.1, it is possible to compile Java incrementally. See the api:org.gradle.api.tasks.compile.JavaCompile[] task for information on how to enable it.
 
-Main goals for incremental compilations are:
+The main goals for incremental compilations are:
 
-* Avoid wasting time compiling source classes that don't have to be compiled. This means faster builds, especially when a change to a source class or a jar does not incur recompilation of many source classes that depend on the changed input.
+* Make incremental builds much faster.
 * Change as few output classes as possible. Classes that don't need to be recompiled remain unchanged in the output directory. An example scenario when this is really useful is using JRebel - the fewer output classes are changed the quicker the JVM can use refreshed classes.
 
 The incremental compilation at a high level:
 
-* The stale class detection favors reliability over speed. The algorithm uses bytecode analysis and deals gracefully with compiler optimizations (inlining of non-private constants), transitive class dependencies, etc. Example: When a class with a public constant changes, we eagerly compile classes that use the same constants to avoid problems with constants inlined by the compiler.
-* To make incremental compilation fast, we cache class analysis results and jar snapshots. The initial incremental compilation can be slower due to the cold caches.
+* A class is recompiled if its source has been changed or if one of its transitive dependencies has been changed.
+* A classes dependencies are determined from type references in its bytecode.
+* Since constants can be inlined, any change to a constant will result in full recompilation. Try to minimize use of constants and replace them with static methods where possible.
+* Since source retention annotations are not visible in bytecode, changes to a source retention annotation will result in full recompilation.
+* Gradle will even incrementally analyze JAR contents, so the incremental compiler works across project boundaries.
+* A loosely coupled code base will have shorter dependency chains and thus compile faster on incremental changes.
```

Comment:
I didn't want to get too technical here, but rather encourage people to follow good software engineering practices, which automatically lead to less dependencies between classes and thus less recompilation. A common practice in Java would be the use of interfaces to define contracts instead of depending on concrete classes.

Reply:
OK. I don't think the detail matters so much here, and folks can do a Google search if they want more information. It might make sense to insert a simple example, such as your example of using interfaces to define contracts rather than concrete classes. That would provide a little bit of context to give folks an idea of what it's about.